rust-src/curve_arithmetic/src/bls12_381_g1hash.rs

```diff
@@ -496,41 +485,139 @@ pub(crate) const K4: [[u64; 6]; 16] = [
     [0x1, 0x0, 0x0, 0x0, 0x0, 0x0],
 ];
 
-pub fn hash_to_g1(bytes: &[u8]) -> G1 {
-    // Concatenate the message with 0u8 and 1u8 respectively
-    // The paper suggests concatenating a single bit - but since the point is to
-    // get two unrelated field elements, concatenating with 0u8 and 1u8 is ok
-    // Instead of forming two new byte arrays we pass a boolean to hash_bytes_to_fq
-    // function below.
-
-    // Notice, this hashing functions varies from the one used by the paper We
-    // use sha512 iteratively until the first 47 bytes + 5 bits of the resulting
-    // digest represents a field element
-    let t0 = hash_bytes_to_fq(false, bytes);
-    let t1 = hash_bytes_to_fq(true, bytes);
-
-    // compute the two points on E1'(Fq) - the 11 isogenous curve
-    let (x0, y0, z0) = simplified_swu(t0);
-    let (x1, y1, z1) = simplified_swu(t1);
-
-    // evaluate the 11-isogeny on the points.
-    // it is faster to add the two points before evaluating the isogeny, but
-    // we omit this for now, since we don't have an efficient implementation
-    // of the curve.
-    let (x0, y0, z0) = iso_11(x0, y0, z0);
-    let (x1, y1, z1) = iso_11(x1, y1, z1);
-
-    // convert into points on the curve
-    let mut p0 = from_coordinates_unchecked(x0, y0, z0).expect("should not happen");
-    let p1 = from_coordinates_unchecked(x1, y1, z1).expect("should not happen");
-
-    // add the two points on E1: y^2 = x^3 + 4
-    p0.add_assign(&p1);
-
-    // Clear cofactors (ensuring that the point is in the correct order subgroup of
-    // the curve)
-    p0.mul_assign(1 + 15_132_376_222_941_642_752);
-    p0
+/// Implements https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-10#section-3
+/// It follows the steps
+///    1. u = hash_to_field(msg, 2)
+///    2. Q0 = map_to_curve(u[0])
+///    3. Q1 = map_to_curve(u[1])
+///    4. R = Q0 + Q1              
+///    5. P = clear_cofactor(R) = h_eff * R   # Clearing cofactor
+///    6. return P,
+/// where the choices of hash_to_field, map_to_curve and h_eff are as described in https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-10#section-8.8.1.
+pub fn hash_to_curve(msg: &[u8]) -> G1 {
+    let dst = "QUUX-V01-CS02-with-BLS12381G1_XMD:SHA-256_SSWU_RO_".as_bytes();
```

Comment:
If you just need the bytes I'd recommend using `b"QUUX-V01-CS02-with-BLS12381G1_XMD:SHA-256_SSWU_RO_"`.

Is the final underscore intended?

Reply:
Regarding the underscore, the test(s) will fail without it, since the testvectors from https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-10#appendix-J.9.1 are computed using the underscore.

Reply:
Good, then it is intended :)

Reply:
Hardcoding the domain separation string from the test case was intended as a quick way of testing without changing any code relying on hash_to_curve. I think we should consider changing it to a parameter and replace QUUX with concordium or similar. (Or something specific to each context if it used in more than one.)

Reply:
I see, that does indeed make sense. Would you propose some string and parametrize the function instead of hardcoding.

In tests we can then use this one.